### PR TITLE
feat(eubfr-cli): bootstrap cli utility - EUBFR-176

### PR DESCRIPTION
**Pull request by kalinchernev**
# PR description

The approach in mind is explained in [EUBFR-176](https://webgate.ec.europa.eu/CITnet/jira/browse/EUBFR-176) ticket. In short, the CLI is making use of the smaller APIs as explained in the [documentation](https://github.com/ec-europa/eubfr-data-lake/blob/master/docs/PUSHING_DATA.md).

As a prerequisite: deploy the infrastructure: `yarn deploy`. This is needed for both resources' creation, but also env vars exports which are used by the CLI

To test the CLI: run `npx eubfr-cli` to see help and follow along.

For example:

On upload:
- upload one file with `npx eubfr-cli upload .content/agri/agri_history.csv -p agri`
- upload several files with `npx eubfr-cli upload .content/inforegio/EUBFR_VIEW_16052018.xml .content/inforegio/regio_projects.json -p inforegio`
- upload all files with `npx eubfr-cli upload`

On show:
- `npx eubfr-cli show` or `npx eubfr-cli show -p inforegio`

On delete:
- delete one or multiple files with `npx eubfr-cli delete agri/somehash`
- delete all files with `npx eubfr-cli delete`

## QA Checklist

When you add a new ETL/producer, please check for the following:

* [ ] Producer's secrets are stored safely
* [ ] Ensure the ETL is added to the corresponding `scripts/` for automated deployment and deletion 
* [ ] There is at least 1 unit test with a jest snapshot for the transform function of the ETL
* [ ] Update the file PRODUCERS_DATA_AVAILABILITY_GRID.md indicating which fields are available in source data of ETL
* [ ] Ensure there is (flow/jsdocs) documentation in the `tranform.js` file which is to be used for automated documentation
* [ ] Generate the necessary documentation pages for the new ETL by `yarn docs:md`

### CODE REVIEW

File: scripts/cleaning/delete-dotfiles.sh [hunk 1]
```diff
@@ -1,9 +1,9 @@
 #!/bin/sh
 
 # Remove old deployment states
-rm -rf **/.serverless
+find . -name "*.serverless" -type d -exec rm -rf {} +
```
Comment (yhuard) [inline scripts/cleaning/delete-dotfiles.sh]: why the change?

File: scripts/content/delete.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env node
+
+// Dependencies
+const shell = require('shelljs'); // eslint-disable-line import/no-extraneous-dependencies
+const config = require('../../config'); // eslint-disable-line import/no-unresolved
+
+Object.keys(config.demo).forEach(producer => {
```
Comment (yhuard) [inline scripts/content/delete.js]: shouldn't the logic be integrated within the cli instead of in a custom script? just wondering

File: tools/eubfr-cli/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+## Usage
+
+```sh
+$ npx eubfr-cli
+```
+
+If you want to make use of the CLI to automatically upload or delete all content of a given stage, you can optionally create a `.content` folder in the root of your project, with the following example structure:
+
+```
+.
```
Comment (yhuard) [inline tools/eubfr-cli/README.md]: that's great, I already had the same file structure 🎉

**Review by yhuard — COMMENTED** [review]
some small remarks, will test later

File: scripts/cleaning/delete-dotfiles.sh (hunk 1 above)
Comment (kalinchernev) [reply scripts/cleaning/delete-dotfiles.sh]: For some unknown reason, it was not working for me in current os. (manjaro kde)
Previously it was, but yesterday it took me an hour to understand why env vars were not correct even after delete and re-deploy.

File: scripts/content/delete.js (hunk 2 above)
Comment (kalinchernev) [reply scripts/content/delete.js]: The logic is indeed in the CLI's delete command. This script is simply iterating through all producers and runs the CLI's logic for each of them.

Comment (yhuard) [reply scripts/content/delete.js]: I mean, instead of learning a new command (`yarn content:delete`), wouldn't it be easier to just run `npx eubfr-cli delete --all`?

File: scripts/cleaning/delete-dotfiles.sh (hunk 1 above)
Comment (yhuard) [reply scripts/cleaning/delete-dotfiles.sh]: that's weird :/

File: tools/eubfr-cli/bin/cli.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  .action((file, options) => {
+    const producer = options.producer || 'agri';
+
+    show({ file, producer });
+  });
+
+program
+  .command('delete [files...]')
+  .description('Deletes files by computedKey. Pass --confirm to delete all.')
+  .option('-p, --producer [producer]', "Producer's name. Defaults to 'agri'.")
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: If the `computed_key` is `inforegio/something`, and nothing is passed as `--producer` (-> it defaults to "agri"), the file is still deleted, right? It seems a bit counterintuitive

**Review by yhuard — CHANGES_REQUESTED** [review]
1. When doing `yarn content:upload`, please filter out all the `.something` files (like `.DS_Store`)
2. `yarn content:upload` crashes if a producer's directory does not exist

File: scripts/content/delete.js (hunk 2 above)
Comment (kalinchernev) [reply scripts/content/delete.js]: The `--all` flag is currently scoped to a producer, if `--producer` is not provided, it defaults to agri. I could default it to all producers?

File: tools/eubfr-cli/bin/cli.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    uploadFiles({ files, credentials });
+  });
+
+program
+  .command('show [file]')
+  .description('Displays files of a given producer.')
+  .option('-p, --producer [producer]', "Producer's name. Defaults to 'agri'.")
+  .action((file, options) => {
+    const producer = options.producer || 'agri';
```
Comment (emeryro) [inline tools/eubfr-cli/bin/cli.js]: Why setting the default producer to agri?
This may cause unwanted political questions.

We may check if the producer is empty and trigger a warning message.

File: tools/eubfr-cli/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+│   ├── EUBFR_VIEW_16052018.xml
+│   └── regio_projects.json
+├── valor
+│   └── valor_sample.xls
+└── wifi4eu
+    └── wifi4euRegistrations.xlsx
+```
+
+Then either run `yarn content:upload` to upload of that content to their respective S3 buckets which trigger the ingestion process for each producer, or run `yarn content:delete` to delete all the currently uploaded content of a given stage.
+
```
Comment (emeryro) [inline tools/eubfr-cli/README.md]: Shouldn't we add the few example you wrote in PR's description here?
It helps to understand how to use the CLI.

File: tools/eubfr-cli/bin/cli.js (hunk 5 above)
Comment (kalinchernev) [reply tools/eubfr-cli/bin/cli.js]: It's our first ETL and it's also first alphabetically :)
I agree it's not too good of an argument to keep it though, will add a warning message.

File: tools/eubfr-cli/README.md (hunk 6 above)
Comment (kalinchernev) [reply tools/eubfr-cli/README.md]: Sure, I'll include further details.

File: services/storage/deleter/.gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# See https://help.github.com/ignore-files/ for more about ignoring files.
+
+# env files
+.env
```
Comment (yhuard) [inline services/storage/deleter/.gitignore]: `.env` files should be ignored globally I think

File: tools/eubfr-cli/bin/cli.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+const showFile = require('../commands/show');
+const deleteFiles = require('../commands/delete');
+
+const missingRequiredInput = '\n error: Missing required input parameters';
+
+// If -p is passed without actual value, it will be boolean true
+// Which is useless information in our case
+const hasValidProducer = options =>
+  options.producer &&
+  options.producer !== '' &&
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: `options.producer !== ''` is useless since you already check `options.producer`

File: tools/eubfr-cli/commands/delete.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      await request.get(aws4.sign(params, creds));
+      return console.log(`${computedKey} has been deleted`);
+    } catch (e) {
+      return console.error(e);
+    }
+  };
+
+  // Marker to delete all files for all producers
+  if (files.length === 0) {
+    return credentials.forEach(async producer => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: Beware of async + forEach!
https://codeburst.io/javascript-async-await-with-foreach-b6ba62bbf404

File: tools/eubfr-cli/commands/delete.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      const producerName = Object.keys(producer)[0];
+      const creds = producer[producerName];
+
+      // Get all the files uploaded by the given producer
+      const response = await client.search({
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: No try/catch around this?

File: tools/eubfr-cli/commands/show.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ * @param {String} file
+ *   The specific file to view. Optional, all items are shown by default.
+ * @param {String} producer
+ *   The producer's name. For example: 'agri', 'budg', etc.
+ */
+const showCommand = async ({ file, producer }) => {
+  let query = '';
+  const index = `${process.env.REACT_APP_STAGE}-meta`;
+  const host = `https://${process.env.REACT_APP_ES_PRIVATE_ENDPOINT}`;
+
```
Comment (yhuard) [inline tools/eubfr-cli/commands/show.js]: You could initialize the query here instead:

```js

File: tools/eubfr-cli/commands/show.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    log: 'warning',
+    apiVersion: '6.2',
+  });
+
+  if (file) {
+    // Show single file
+    query = {
+      index,
+      type: 'file',
+      q: `computed_key:"${file}"`,
```
Comment (yhuard) [inline tools/eubfr-cli/commands/show.js]: And then: 

```js 
query.q = `computed_key:"${file}"`;
```

File: tools/eubfr-cli/commands/show.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    };
+  } else {
+    // Show all files
+    query = {
+      index,
+      type: 'file',
+      body: {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/show.js]: The same with `body`

File: tools/eubfr-cli/commands/upload.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+ * @param {Array} files
+ *   The files to upload.
+ * @param {Array} credentials
+ *   List of credentials for the producers.
+ */
+const uploadCommand = ({ files, credentials }) => {
+  if (!process.env.SIGNED_UPLOADS_API) {
+    return console.error(
+      "SIGNED_UPLOADS_API environment variable is missing. Please redeploy by running 'yarn deploy' from project root"
+    );
```
Comment (yhuard) [inline tools/eubfr-cli/commands/upload.js]: No `process.exit(1)` here?

File: tools/eubfr-cli/commands/upload.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        // Removing double quotes to build a correct path.
+        uri: signedUrl.replace(/["]+/g, ''),
+        body: fs.readFileSync(path.resolve(file)),
+      });
+    } catch (e) {
+      return console.error(e);
+    }
+  };
+
+  return credentials.forEach(async producer => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/upload.js]: `async`, `forEach`, you know the drill ;)

File: tools/eubfr-cli/lib/getProducerFiles.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+const readDir = promisify(fs.readdir);
+
+module.exports = async producer => {
+  let files = [];
+  const dir = path.resolve(`.content/${producer}`);
+  try {
+    if (fs.existsSync(dir)) {
+      const results = await readDir(dir);
+      // Filter hidden/dot files
+      files = results.filter(file => !/(^|\/)\.[^\/\.]/g.test(file)); // eslint-disable-line no-useless-escape
```
Comment (yhuard) [inline tools/eubfr-cli/lib/getProducerFiles.js]: Why did you disable the eslint rule? 😆

**Review by yhuard — CHANGES_REQUESTED** [review]

File: tools/eubfr-cli/bin/cli.js [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+      console.error('\n error: Please specificy producer with a name.');
+      process.exit(1);
+    }
+
+    if (producer && file && hasValidProducer(options)) {
+      console.error('\n error: Pass file or producer only, not both.');
+      process.exit(1);
+    }
+
+    showFile({ file, producer });
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: `showFile` is async but the current function isn't

File: tools/eubfr-cli/commands/delete.js [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        } catch (e) {
+          return console.error(e);
+        }
+      })
+    );
+  }
+
+  // In case files are specified, delete them.
+  Promise.all(
+    files.map(file => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: missing `async` keyword

File: tools/eubfr-cli/bin/cli.js [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+  });
+
+program
+  .command('delete [files...]')
+  .description('Deletes files by computed_key field.')
+  .option('-c, --confirm [confirm]', 'Flag certainty of an operation.')
+  .action((files, options) => {
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: missing `async` keyword

File: tools/eubfr-cli/bin/cli.js [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    const producers = getAllProducers();
+    const credentials = producers.map(producer => ({
+      [producer]: getCredentials(producer),
+    }));
+
+    if (options.confirm) {
+      deleteFiles({ files, credentials });
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: better with `await`

File: tools/eubfr-cli/bin/cli.js [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
+      // Initiate the prompt interface.
+      const rl = readline.createInterface({
+        input: process.stdin,
+        output: process.stdout,
+      });
+
+      rl.question('Are you sure? <yes|y> ', answer => {
+        if (answer === 'y' || answer === 'yes') {
+          deleteFiles({ files, credentials });
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: better with `await`

File: tools/eubfr-cli/commands/delete.js [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+              },
+            },
+          });
+
+          const allFiles =
+            response.hits && response.hits.hits
+              ? response.hits.hits.map(file => file._source.computed_key)
+              : [];
+
+          return allFiles.map(computedKey => deleteFile(computedKey, creds));
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: `return Promise.all(allFiles.map(async computedKey => deleteFile(computedKey, creds)));`, isn't it?

File: tools/eubfr-cli/lib/extractCredentials.js [hunk 23]
```diff
@@ -0,0 +1,11 @@
+const extractCredentials = ({ producerKey, credentials }) => {
+  let creds = '';
+  credentials.forEach(producerCredentials => {
```
Comment (yhuard) [inline tools/eubfr-cli/lib/extractCredentials.js]: what about using `Array.find` instead?

File: tools/eubfr-cli/lib/getProducerFiles.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+const { promisify } = require('util');
+
+const readDir = promisify(fs.readdir);
+
+module.exports = async producer => {
+  let files = [];
+  const dir = path.resolve(`.content/${producer}`);
+  try {
+    if (fs.existsSync(dir)) {
+      const results = await readDir(dir);
```
Comment (yhuard) [inline tools/eubfr-cli/lib/getProducerFiles.js]: why not using https://nodejs.org/api/fs.html#fs_fs_readdirsync_path_options?

**Review by yhuard — CHANGES_REQUESTED** [review]
- so many changes in yarn.lock, seems strange to me
- some small confusions with async/await

File: tools/eubfr-cli/lib/extractCredentials.js (hunk 23 above)
Comment (kalinchernev) [reply tools/eubfr-cli/lib/extractCredentials.js]: There are several methods to solve this, the question is what is the benefit to change to Array.find when it's readable and it works?

Comment (yhuard) [reply tools/eubfr-cli/lib/extractCredentials.js]:
```

File: tools/eubfr-cli/commands/upload.js [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+
+      return console.timeEnd(`${fileName} has been uploaded`);
+    } catch (e) {
+      return console.error(
+        `Upload of ${fileName} failed with ${e.statusCode} code.`
+      );
+    }
+  };
+
+  credentials.map(producer => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/upload.js]: here you could return `return Promise.all(credentials.map(producer => {`, thus keeping the control over the process

File: tools/eubfr-cli/bin/cli.js [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (!producerIsSet && !filesAreSet) {
+      console.log('All files for all producers will be uploaded.');
+      const producers = getAllProducers();
+      credentials = producers.map(producer => ({
+        [producer]: getCredentials(producer),
+      }));
+    }
+
+    uploadFiles({ files, credentials });
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: it'd be better to await it (see comment later on `upload.js`)

File: tools/eubfr-cli/commands/delete.js [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+const path = require('path');
+
+const aws4 = require('aws4');
+const dotenv = require('dotenv');
+const elasticsearch = require('elasticsearch');
+const request = require('request-promise-native');
+
+const getServiceLocation = require('../lib/getServiceLocation');
+
+dotenv.config({
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: `dotenv` should only be called in the main `cli.js` file. Then, you could write the commands as pure functions 😉

File: tools/eubfr-cli/commands/delete.js [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+  const host = `https://${process.env.REACT_APP_ES_PRIVATE_ENDPOINT}`;
+
+  const client = elasticsearch.Client({
+    host,
+    log: 'warning',
+    apiVersion: '6.2',
+  });
+
+  // Reuse throughout the two delete approaches.
+  const deleteFile = async (computedKey, creds) => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: it would be cleaner to define the function outside of the parent function, unless it's absolutely necessary

File: tools/eubfr-cli/commands/delete.js [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Deletes files from S3 for a given producer.
+ *
+ * @param {Array} files
+ *   List of files to delete.
+ * @param {Array} credentials
+ *   List of credentials for the producers.
+ */
+const deleteCommand = async ({ files, credentials }) => {
```
Comment (yhuard) [inline tools/eubfr-cli/commands/delete.js]: Technically, this `async` is not needed, the function doesn't directly uses `await` and already returns a Promise but it's okay to keep it, it's like a safety net that ensures we always return a Promise :)

**Review by yhuard — CHANGES_REQUESTED** [review]
a few more comments, the most important being the one about dotenv

File: tools/eubfr-cli/bin/cli.js (hunk 26 above)
Comment (kalinchernev) [reply tools/eubfr-cli/bin/cli.js]: I don't think so, the function is not async. Why awaiting it if not async?

File: tools/eubfr-cli/bin/cli.js [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+const uploadFiles = require('../commands/upload');
+const showFile = require('../commands/show');
+const deleteFiles = require('../commands/delete');
+
+// If -p is passed without actual value, it will be boolean true
+// Which is useless information in our case
+const hasValidProducer = options =>
+  options.producer && typeof options.producer !== 'boolean';
+
+const endpoints = getEndpoints();
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: it should maybe be wrapped in a try/catch block (in case the `.env` files don't exist)

File: tools/eubfr-cli/bin/cli.js [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (!producerIsSet && !filesAreSet) {
+      console.log('All files for all producers will be uploaded.');
+      const producers = getAllProducers();
+      credentials = producers.map(producer => ({
+        [producer]: getCredentials(producer),
+      }));
+    }
+
+    uploadFiles({ files, credentials, endpoints });
```
Comment (yhuard) [inline tools/eubfr-cli/bin/cli.js]: I maintain it would be better to `await` this instead of just letting the Promise resolve in the background 😉

**Review by yhuard — CHANGES_REQUESTED** [review]
almost there :)

File: tools/eubfr-cli/bin/cli.js (hunk 30 above)
Comment (kalinchernev) [reply tools/eubfr-cli/bin/cli.js]: If there are no .env files and this function does not manage to gather information it will return `{}` and the ensureVariables will warn and exit for the first missing required environment variable.

File: tools/eubfr-cli/lib/getEndpoints.js [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+  'storage-signed-uploads',
+  'demo-dashboard-client',
+];
+
+const getEndpoints = () => {
+  const endpoints = {};
+
+  services.forEach(service => {
+    const dotEnvFile = path.resolve(`${getServiceLocation(service)}/.env`);
+    const vars = envfile.parseFileSync(dotEnvFile);
```
[the remark this answers is not in the document]
Comment (kalinchernev) [reply tools/eubfr-cli/lib/getEndpoints.js]: It returns an empty object, thus leaving `endpoints` defaulting to `{}` to `{}`.

**Review by yhuard — APPROVED** [review]
